### PR TITLE
Polars: OnceFrame

### PR DESCRIPTION
**Pull request by Shoeboxam**
Currently in Polars, Measurements emit LazyFrames. LazyFrames can be repeatedly executed, and each time it is executed, the privacy loss increases (but is not tracked anywhere). 

This PR instead returns a queryable with the LazyFrame in its internal state. When the user wants to execute the LazyFrame, the queryable executes the LazyFrame and then frees it, making it impossible to execute again.

A queryable is used so that the user can control how the LazyFrame is executed: either collect into memory, sink to a CSV, sink to Parquet, to IPC or to JSON. Queryables may also share the query plan, or allow the user to configure query plan optimizations.

In both the Python and Rust APIs, the Queryable emitted mimics a (very) limited subset of the LazyFrame API. We can expand this API in the future as needed.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1602?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1602** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1602?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1597** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1597?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1596** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1596?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1592** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1592?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1591** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1591?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1600** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1600?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1590** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1542** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1503](https://github.com/opendp/opendp/pull/1503) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1503?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mccalluc** [comment]
I'll start a more thorough code review, but just skimming it, it's hard for me to see exactly what this means for the user. Would it be possible to update or copy
- #1632

and integrate that example into this PR to demonstrate the interface?

I also have some small PRs out that improve test coverage, particularly on generated code. Having those checks in place would make it easier for me to see what code here is actually being exercised.

Finally, I'd worry about holding up the release for this. When we do make a tagged release with this, I think we also need at least orphan docs for the new functionality.

For the code review, I'll try putting in more a lot more comments along the way. I'm not sure how this will work, but it might be good for me to be really transparent about what I understand and what I don't understand.

File: python/src/opendp/_data.py
Comment (mccalluc) [inline python/src/opendp/_data.py:32]: This is used in turn by `_lazyframe`, but that function doesn't seem to have any usages. Do we either/both of these?

File: python/src/opendp/_data.py
Comment (mccalluc) [inline python/src/opendp/_data.py:308]: If we have a method to return a DataFrame, should that be enough? Is there a concern that going through an intermediate data structure will be too slow or inefficient for large datasets and we need this method and `onceframe_sink_parquet` to make it more direct? If that is the motivation, worry about premature optimization. What determines that these two formats are the ones we support?

File: python/src/opendp/mod.py
Comment (mccalluc) [inline python/src/opendp/mod.py:432]: Oh! It looks like you had a wrapper around `RuntimeType.parse` before, and are moving away from that. There's probably a good reason, but I don't understand the motivation.

File: python/src/opendp/polars.py
Comment (mccalluc) [inline python/src/opendp/polars.py:154]: Noted above, but I think this is unused?

File: python/src/opendp/prelude.py
Comment (mccalluc) [inline python/src/opendp/prelude.py:50]: I don't think any of the python code uses this, right now, so could leave it out of this PR? At one point, I think we were thinking of including this method and others under a new `dp.x` namespace, which would have all the methods which require optional installs?

File: python/test/test_polars.py
Comment (mccalluc) [inline python/test/test_polars.py:121]: Is it wrong to `sort` as a post processing step? Or maybe it just exercises more of the machinery to do it here? If it's not a good thing to do, that should get mentioned in the docs. (Though maybe it's already there?)

File: rust/src/data/ffi/mod.rs
Comment (mccalluc) [inline rust/src/data/ffi/mod.rs:15]: Reminder to self: `::polars` resolves ambiguity -- It specifies the external crate instead of the internal module.

**Review by mccalluc — COMMENTED** [review]
Just some comments on the Python code so far. Not sure that all the functions that get exposed need to be exist, but you know the big picture better than me.

Later this afternoon I'll have another batch of comments on the Rust code.

**Comment by Shoeboxam** [comment]
This PR makes the user experience worse-- users can only conduct a limited set of actions on a OnceFrame, instead of all possible actions with LazyFrame. However it prevents users from violating privacy guarantees by executing the plan multiple times. This should be merged before releasing 0.10 to ensure that the privacy guarantees we give are valid.

File: python/src/opendp/_convert.py
Comment (mccalluc) [reply python/src/opendp/_convert.py:150]: (I now see the change in `mod.py` below, so it looks like it used to be like that, so it's probably changing for a good reason?)

File: R/opendp/src/opendp.h
Comment (Shoeboxam) [reply R/opendp/src/opendp.h:740]: Yes, this should be blacklisted.

File: python/src/opendp/mod.py
Comment (Shoeboxam) [reply python/src/opendp/mod.py:432]: The query type of a queryable will never change (it is static), so there's no need to repeatedly retrieve the query type at every call. Since I already need to know the query type to convey it into the correct Python type (Queryable or OnceFrame), I might as well simply the property to just an attribute.

File: python/src/opendp/polars.py
Comment (Shoeboxam) [reply python/src/opendp/polars.py:154]: I adjusted this to make it public, and to assert "honest-but-curious".

File: python/src/opendp/_data.py
Comment (Shoeboxam) [reply python/src/opendp/_data.py:308]: This is a great comment. That is a motivation, yes, but I also had another hope for OnceFrames that are not yet used by this PR. I think `trait PrivateLogicalPlan` shouldn't emit measurements with type `LogicalPlan`, as this also has the same problem as the `LazyFrame` type. `OnceFrame` better suits this. I was thinking this would be in a future PR.

Arguably we could just make `OnceFrame` internal, and only release data frames from `make_private_lazyframe`. However, grouping key sets can get larger than memory, so permitting these use-cases would be useful. One minor simplification here is that I could remove the auxiliary `sink_*` constructors, and just keep `collect` and the honest-but-curious `lazyframe`.

File: python/test/test_polars.py
Comment (Shoeboxam) [reply python/test/test_polars.py:121]: The output of the measurement is a OnceFrame now, which has a more limited API. One of the additions in this PR is to allow sorting within `make_private_lazyframe`. Operations that are on what are now OnceFrames have been moved into the constructor.

**Review by mccalluc — COMMENTED** [review]
Overall looks good. Many of the questions here parallel questions from the python review. Talk and decide on next steps?

File: rust/src/measurements/make_private_expr/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/make_private_expr/mod.rs:36]: Note to self: should document why the constructor is honest-but-curious.

Comment (Shoeboxam) [reply rust/src/measurements/make_private_expr/mod.rs:36]: Yes, file an issue from this comment-- maybe this should be a hand-written doc with an automated check?

File: python/src/opendp/polars.py
Comment (Shoeboxam) [reply python/src/opendp/polars.py:152]: I've removed the `sink_*` methods and associated machinery. For the time being, if someone wants it, they should call `.lazy()` with honest-but-curious first.

File: rust/src/measurements/make_private_lazyframe/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/make_private_lazyframe/mod.rs:106]: Maybe I'm being too forward-thinking, but I anticipate the removal of "contrib" to happen piecemeal as each individual PDF proof gets a green light.

Comment (Shoeboxam) [reply rust/src/measurements/make_private_lazyframe/mod.rs:106]: I think this matcher doesn't need its own contrib flag, though. Admittedly somewhat arbitrary.

File: rust/src/data/ffi/polars.rs
Comment (Shoeboxam) [reply rust/src/data/ffi/polars.rs:53]: I narrowed down to just `collect` and `lazy`.

File: python/src/opendp/prelude.py
Comment (Shoeboxam) [reply python/src/opendp/prelude.py:50]: On second thought, I think this does make at least some sense. I added this because otherwise the plugin namespace won't get registered. You still can't access any imports from that module in the prelude.

File: rust/src/measurements/make_private_expr/mod.rs
Comment (mccalluc) [reply rust/src/measurements/make_private_expr/mod.rs:36]: - Filed #1662

**Review by mccalluc — COMMENTED** [review]
Just a couple questions. After the tests pass, maybe just grab me and see if there are any last questions, and then I'll approve?

**Comment by Shoeboxam** [comment]
I think it's ok that diffs for blacklisted functions are showing up in the R header file, because the generated header file is meant generally for any language for bind with, and not all the functions in a header file need to be used.

File: python/src/opendp/prelude.py
Comment (Shoeboxam) [reply python/src/opendp/prelude.py:53]: This is no longer a problem after the PR restack.

Comment (Shoeboxam) [reply python/src/opendp/prelude.py:53]: The module should be public, it just shouldn't be accessible via `dp.opendp.polars`. That would be such a weird import.

**Review by mccalluc — APPROVED** [review]
Spot checking the earlier threads, looks good. Let me know if there's anything you'd like me to look at more carefully.

Make sure CI passes before merge? Not sure why it wasn't run on last commits.

**Pull request merged**
